qutip/core/dimensions.py

```diff
@@ -343,3 +324,553 @@ def dims_idxs_to_tensor_idxs(dims, indices):
+
+
+def _frozen(*args, **kwargs):
+    raise RuntimeError("Dimension cannot be modified.")
+
+
+class MetaSpace(type):
+    def __call__(cls, *args, rep=None):
+        """
+        Select which subclass is instanciated.
+        """
+        if cls is Space and len(args) == 1 and isinstance(args[0], list):
+            # From a list of int.
+            return cls.from_list(*args, rep=rep)
+        elif len(args) == 1 and isinstance(args[0], Space):
+            # Already a Space
+            return args[0]
+
+        if cls is Space:
+            if len(args) == 0:
+                # Empty space: a Field.
+                cls = Field
+            elif len(args) == 1 and args[0] == 1:
+                # Space(1): a Field.
+                cls = Field
+            elif len(args) == 1 and isinstance(args[0], Dimensions):
+                # Making a Space out of a Dimensions object: Super Operator.
+                cls = SuperSpace
+            elif len(args) > 1 and all(isinstance(arg, Space) for arg in args):
+                # list of space: tensor product space.
+                cls = Compound
+
+        if settings.core['auto_tidyup_dims']:
+            if cls is Compound and all(isinstance(arg, Field) for arg in args):
+                cls = Field
+            if cls is SuperSpace and args[0].type == "scalar":
+                cls = Field
+
+        args = tuple([
+            tuple(arg) if isinstance(arg, list) else arg
+            for arg in args
+        ])
+
+        if cls is Field:
+            return cls.field_instance
+
+        if cls is SuperSpace:
+            args = *args, rep or 'super'
+
+        if args not in cls._stored_dims:
+            instance = cls.__new__(cls)
+            instance.__init__(*args)
+            cls._stored_dims[args] = instance
+        return cls._stored_dims[args]
+
+    def from_list(cls, list_dims, rep=None):
+        if not isinstance(list_dims[0], list):
+            # Tensor
+            spaces = [Space(size) for size in list_dims]
+        elif len(list_dims) == 1:
+            # [[2, 3]]: tensor with an extra layer of list.
+            spaces = [Space(size) for size in list_dims[0]]
+        elif len(list_dims) % 2 == 0:
+            # Superoperators or tensor of Superoperators
+            spaces = [
+                Space(Dimensions(
+                    Space(list_dims[i+1]),
+                    Space(list_dims[i])
+                ), rep=rep)
+                for i in range(0, len(list_dims), 2)
+            ]
+        else:
+            raise ValueError(f'Format not understood {list_dims}')
+
+        if len(spaces) == 1:
+            return spaces[0]
+        elif len(spaces) >= 2:
+            return Space(*spaces)
+        raise ValueError("Bad list format")
+
+
+class Space(metaclass=MetaSpace):
+    _stored_dims = {}
+    def __init__(self, dims):
+        idims = int(dims)
+        if dims <= 0 or idims != dims:
+            raise ValueError("Dimensions must be integers >= 0")
```

Comment:
```suggestion
            raise ValueError("Dimensions must be integers > 0")
```
Otherwise this doesn't match the logic in the `if` above it.